---
title: "Lecture 37: Accessing GEO directly, Design Matrices, and Differential Expression for Arrays, Reading: the vignette for limma"
author: "Miles Pufall, May 1, 2020"
output: 
  powerpoint_presentation:
    reference_doc: "/Users/milespufall/Dropbox/miles/teaching/LDS/lecture_images/lecture_23_template.potx"
always_allow_html: true
---

```{r}
# install.packages("rhdf5")
# library("rhdf5")
library("GenomicFeatures")
library("vsn")
library("pheatmap")
library("RColorBrewer")
library("PoiClaClu")
library("apeglm")
library("genefilter")
library("AnnotationDbi")
library("org.Hs.eg.db")
library("ReportingTools")
library("readr")
library("tidyverse")
library("viridis")
library("qvalue")
library("ggpubr")
library("limma")
library("DESeq2")
library("GEOquery")
library("readxl")
library("writexl")
library("openxlsx")

```


```{r read_from_GEO, message = FALSE, warnings = FALSE, error = FALSE}
# load series and platform data from GEO
gset <- getGEO("GSE7007", GSEMatrix =TRUE, AnnotGPL=TRUE)
if (length(gset) > 1) idx <- grep("GPL96", attr(gset, "names")) else idx <- 1
gset <- gset[[idx]]

# make proper column names to match toptable 
fvarLabels(gset) <- make.names(fvarLabels(gset))

# group membership for all samples
gsms <- "X44552233001XXXXXXXXXXXXXXXXXXXXXXXXXXX"
sml <- strsplit(gsms, split="")[[1]]

# filter out excluded samples (marked as "X")
sel <- which(sml != "X")
sml <- sml[sel]
gset <- gset[ ,sel]

## Getting just sample, expression, probe data from GSE
ex <- exprs(gset)

# log2 transformation
qx <- as.numeric(quantile(ex, c(0., 0.25, 0.5, 0.75, 0.99, 1.0), na.rm=T))
LogC <- (qx[5] > 100) ||
  (qx[6]-qx[1] > 50 && qx[2] > 0)
if (LogC) { ex[which(ex <= 0)] <- NaN
exprs(gset) <- log2(ex) }

## Checking to see if data has been normalized
#plotDensities(ex, legend = FALSE)

## It has not, so now we normalize with normalizeBetweenArrays
#norm_ex <- normalizeBetweenArrays(ex)

## Density plots shows data normalized
#plotDensities(norm_ex, legend = FALSE)

## Extract sample annotations
pdata <- pData(gset)

## Rename columns norm_ex to sample ids (from "GSMXXXX" to "A673_shCT," etc.)
#colnames(norm_ex) <- pdata$title
```

``` {r}
## Cluster
#mat_dist <- as.matrix(dist(t(norm_ex)))

## Heatmap of cluster
#pheatmap(mat_dist, color = blues9)
```


```{r PCA plot}
#pca_samples <- prcomp(t(norm_ex), scale. = TRUE)

#pca_samples.var <- pca_samples$sdev^2
#Calculate the % contribution of each component

#pca_samples.var.per <- round(pca_samples.var/sum(pca_samples.var)*100, 1)

#pca.data <- data.frame(Sample=rownames(pca_samples$x),  X=pca_samples$x[,1],  Y=pca_samples$x[,2])

#ggplot(data=pca.data, aes(x=X, y=Y, label = Sample)) +
 # geom_text() +  xlab(paste("PC1 - ", pca_samples.var.per[1], "%", sep="")) +  
 # ylab(paste("PC2 - ", pca_samples.var.per[2], "%", sep="")) +
# theme_bw()
```

All we need for our design matrix is the sample names and the treatments.

Let's use dplyr to make this:

```{r}
## Subset conditions from pdata
cond <- pdata %>%
  dplyr::select(title, geo_accession) %>%
  mutate_all(~replace(., is.na(.), 0)) %>%
  dplyr::mutate(title = stringr::str_replace_all(string = title,
                                               pattern = "_rep1", replacement = "")) %>%
  dplyr::mutate(title = stringr::str_replace_all(string = title,
                                               pattern = "_rep2", replacement = "")) %>%
  as.data.frame()

## Making design matrix
cond$title <- factor(cond$title, levels=c("EW24_shCTL", "EW24_shEF1","SKNMC_shCTL","SKNMC_shEF1", "A673_siCT", "A673_siEF1"))

rownames(cond) <- cond$geo_accession

model <- model.matrix(~0 + title, cond)
model

fit <- lmFit(gset, model)
head(coef(fit))
colnames(fit)

annot <- fit$genes

annot_tidy <- annot %>%
  dplyr::select(1:3)

```

```{r}
## isolating AP-1 protein expression values
express_value <- as.data.frame(coef(fit))

colnames(express_value) <- c("EW24_shCtl","EW24_shEF1","SKNMC_shCtl","SKNMC_shEF1","A673_siCtl","A673_siEF1")

express_value$probe <- rownames(express_value)

ex_value_annot <- inner_join(express_value, annot_tidy, 
  by = join_by("probe" == "ID"))

rownames(ex_value_annot) <- ex_value_annot$probe

ex_value_ap1 <- inner_join(ex_value_annot, ap1_genes, 
  by = join_by("Gene.symbol" == "ap1_genes")) %>%
  unique()
```


Then, we can make the contrasts between conditions that we'd like, and perform our differential expression test using the mean expressions.

```{r}
cont.matrix <- makeContrasts(EW24 = titleEW24_shEF1 - titleEW24_shCTL, SKNMC = titleSKNMC_shEF1 - titleSKNMC_shCTL, A673 = titleA673_siEF1 - titleA673_siCT, levels=model)

fit2 <- contrasts.fit(fit, cont.matrix)

# head(coef(fit2))
```

## Perform the statistical comparisons

Then we can do our statistical tests, using eBayes()

```{r}
fit2 <- eBayes(fit2)

# topTable(fit2)
```


Yet another new class!

## Extracting and storing the differential expression table

There is a convenient function called topTable() which will give a summary of the results.

It selects results based on an ANOVA of whether there was ANY change under any contrasts:

```{r}
## Subsetting individual cell line differentially regulated genes (p >= 0.05)
# LOGFC ACTUALLY MEANS LOG2FC (SEE LIMMA EBAYES NOTES)
all.A673 <- topTable(fit2, coef = "A673",number = Inf, adjust.method = "fdr")
all.A673.tidy <- all.A673 %>%
  dplyr::select(1:3,22,26) %>%
  group_by(Gene.symbol) %>%
  dplyr::filter(adj.P.Val == min(adj.P.Val))

colnames(all.A673.tidy) <- c("probe","description","symbol","log2FC","adj.p.val")

deg.A673 <- dplyr::filter(all.A673.tidy, adj.p.val <= 0.05)
upreg.A673 <- dplyr::filter(deg.A673, log2FC >= 1)

###

all.SKNMC <- topTable(fit2, coef = "SKNMC",number = Inf, adjust.method = "fdr")
all.SKNMC.tidy <- all.SKNMC %>%
  dplyr::select(1:3,22,26)%>%
  group_by(Gene.symbol) %>%
  dplyr::filter(adj.P.Val == min(adj.P.Val))

colnames(all.SKNMC.tidy) <- c("probe","description","symbol","log2FC","adj.p.val")

deg.SKNMC <- dplyr::filter(all.SKNMC.tidy, adj.p.val <= 0.05)
upreg.SKNMC <- dplyr::filter(deg.SKNMC, log2FC >= 1)
##

all.EW24 <- topTable(fit2, coef = "EW24",number = Inf, adjust.method = "fdr")
all.EW24.tidy <- all.EW24 %>%
  dplyr::select(1:3,22,26) %>%
  group_by(Gene.symbol) %>%
  dplyr::filter(adj.P.Val == min(adj.P.Val))

colnames(all.EW24.tidy) <- c("probe","description","symbol","log2FC","adj.p.val")

deg.EW24 <- dplyr::filter(all.EW24.tidy, adj.p.val <= 0.05)

upreg.EW24 <- dplyr::filter(deg.EW24, log2FC >= 1)


```

```{r}
ap1_genes <- c("ATF1","ATF2","ATF3","ATF4","ATF5","ATF6","ATF6B","ATF7","BATF","BATF2","BATF3","FOS","FOSB","FOSL1","FOSL2","JDP2","JUN","JUNB","JUND","MAF","MAFA","MAFB","MAFF","MAFG","MAFK","NRL")

ap1_genes <- as.data.frame(ap1_genes)

colnames(ap1_genes) <- c("symbol")

```

```{r}
all.A673.tidier <- all.A673.tidy %>%
                    dplyr::select(3:5)

colnames(all.A673.tidier) <- c("symbol","log2FC.A673","padj.A673")

#rownames(all.A673.tidier) <- all.A673.tidier$symbol
#all.A673.tidier <- dplyr::select(2,3) %>% view()

all.SKNMC.tidier <- all.SKNMC.tidy %>%
                    dplyr::select(3:5)

colnames(all.SKNMC.tidier) <- c("symbol","log2FC.SKNMC","padj.SKNMC")

all.EW24.tidier <- all.EW24.tidy %>%
                    dplyr::select(3:5)

colnames(all.EW24.tidier) <- c("symbol","log2FC.EW24","padj.EW24")
```

```{r}
## Making spreadsheet of all values, unfiltered by padj or log2fc
all_gse7007_list <- list(all.A673.tidier,
                        all.SKNMC.tidier,
                        all.EW24.tidier)

# merge all GSE7007 all data frames together
gse7007_all_table <- all_gse7007_list %>% 
  purrr::reduce(full_join, by='symbol') %>%
  as.data.frame()

library(openxlsx)

openxlsx::write.xlsx(gse7007_all_table,"/Volumes/argon_home/GSE7007/all_gse7007.xlsx")

## Reopening excel file
gse7007_all_table <- read_excel("/Volumes/argon_home/GSE7007/all_gse7007.xlsx")
```

```{r}
## Subsetting unfiltered AP1 values for heatmap
ap1.gse7007 <- inner_join(gse7007_all_table, ap1_genes,
                          by = "symbol")

gse7007_ap1_log2fc <- ap1.gse7007 %>% dplyr::select(1,2,4,6) %>% as.data.frame()

gse7007_ap1_padj <- ap1.gse7007 %>% dplyr::select(1,3,5,7) %>% as.data.frame()

colnames(gse7007_ap1_log2fc) <- c("symbol","A673","SKNMC","EW24")

colnames(gse7007_ap1_padj) <- c("symbol","A673","SKNMC","EW24")

rownames(gse7007_ap1_log2fc) <- gse7007_ap1_log2fc$symbol

rownames(gse7007_ap1_padj) <- gse7007_ap1_padj$symbol


```

```{r}
gse7007_ap1_padj_num <- gse7007_ap1_padj %>% dplyr::select(where(is.numeric))

gse7007_ap1_log2fc_num <- gse7007_ap1_log2fc %>% dplyr::select(where(is.numeric))

## Pheatmaps for AP1 (not filtered by padj or log2fc, just showing all values)
pheatmap(gse7007_ap1_log2fc_num, angle = 45, show_rownames = TRUE, 
         treeheight_row = 0, treeheight_col = 0,cellwidth = 22, cellheight = 11,
         display_numbers = ifelse(gse7007_ap1_padj_num <= 0.05
         & (gse7007_ap1_log2fc_num > 1 | gse7007_ap1_log2fc_num < -1), "*", ""),
  width = 3.5, height = 5,
 main = paste("GSE7007"), 
 filename = "/Volumes/argon_home/GSE7007/GSE7007_AP1_Heatmap_padj-log2fc.pdf")


```


```{r}
## Making spreadsheet of all degs
degs_gse7007 <- list("A673" = deg.A673,
                  "SKNMC" = deg.SKNMC,
                  "EW24" = deg.EW24)

openxlsx::write.xlsx(degs_gse7007,"/Volumes/argon_home/GSE7007/degs_gse7007.xlsx")
```

```{r}
## Making a single list of containing upregulated genes for each cell line
upreg_gse7007 <- list("A673" = upreg.A673,
                  "SKNMC" = upreg.SKNMC,
                  "EW24" = upreg.EW24)

openxlsx::write.xlsx(upreg_gse7007,"/Volumes/argon_home/GSE7007/upreg_gse7007.xlsx")
```

```{r}
GSE7007_ECM_Enrichment <- read_excel("/Volumes/argon_home/GSE7007/GSE7007_ECM_Enrichment.xlsx") %>%
  dplyr::select(1:6) %>%
  as.data.frame()

     

```

```{r}
## Filtering by enrichment for the pathway ECM focal constituent pathway
ecm_structure_gse7007 <- dplyr::filter(GSE7007_ECM_Enrichment, Pathways == "Extracellular matrix structural constituent")

## Plotting focal adhesion Enrichment
ecm_structure_gse7007_plot <- ggplot(ecm_structure_gse7007, aes(x=Cell_Line, y=(log2(Fold_Enrichment)))) +
  geom_segment(aes(x=Cell_Line, xend=Cell_Line, y=0, yend=(log2(Fold_Enrichment))), linewidth=1,color="black") +
  geom_point(aes(size = Genes)) +
  theme_minimal() +
  theme(
    axis.ticks.x = element_blank(),
    legend.text = element_text(size = 25),
    legend.title = element_text(size = 25),
    axis.text.x = element_text(size = 25),
    axis.text.y = element_text(size = 25),
    axis.title.y = element_text(size = 25),
    plot.title = element_text(size = 30, hjust = 0.5, face = "bold"),
    plot.subtitle = element_text(size = 25,hjust = 0.5),
    strip.text = element_text(size = 20),
  ) +
  ggtitle("ECM Structural Constituent", subtitle = "(GO: Molecular Function)") +
  xlab("") +
  ylab("Log2 Fold Enrichment") + 
  scale_size(range = c(3,8))

ggsave(plot = ecm_structure_gse7007_plot, filename="GSE7007_Structure_Plot.pdf",
       path = "/Volumes/argon_home/RNAseq-overlap/overlaps-cell-line-sep", dpi = 300, 
       width = 8, height = 5, units = "in")
```


```{r}

ap1_genes <- c("ATF1","ATF2","ATF3","ATF4","ATF5","ATF6","ATF6B","ATF7","BATF","BATF2","BATF3","FOS","FOSB","FOSL1","FOSL2","JDP2","JUN","JUNB","JUND","MAF","MAFA","MAFB","MAFF","MAFG","MAFK","NRL")

ap1_genes <- as.data.frame(ap1_genes)


ew24_ap1_merge <- inner_join(all.EW24.tidy, ap1_genes, 
  by = join_by("symbol" == "ap1_genes"))

sknmc_ap1_merge <- inner_join(all.SKNMC.tidy, ap1_genes, 
  by = join_by("symbol" == "ap1_genes"))
  
  
ew24_sknmc_merge <- inner_join(ew24_ap1_merge, sknmc_ap1_merge, 
  by = "symbol",
  copy = FALSE,
  suffix = c(".EW24", ".SKNMC"),
  keep = NULL) %>%
  dplyr::select(1:5,8,9) %>%
  as.data.frame()

#colnames(ew24_sknmc_merge)

cols_merge <- c("probe","description","symbol","log2FC.EW24","adj.p.val.EW24","log2FC.SKNMC","adj.p.val.SKNMC")

colnames(ew24_sknmc_merge) <- cols_merge

rownames(ew24_sknmc_merge) <- ew24_sknmc_merge$symbol

```

```{r log2FC AP1 heatmap}
gse7007_ap1_hp_log2FC <- ew24_sknmc_merge %>%
  dplyr::select(4,6)

colnames(gse7007_ap1_hp_log2FC) <- c("EW24","SKNMC")

pheatmap(gse7007_ap1_hp_log2FC, cellwidth = 22, cellheight = 11, treeheight_row = 0, treeheight_col = 0,main = paste("GSE7007"))

```

```{r Hallmark EMT Genes Heatmap}
# Importing list of Hallmark EMT Genes from ShinyGO 0.76
hallmark_emt_genes <- read_excel("/Volumes/argon_home/siEF-proj/hallmark_emt_genes.xlsx") %>% as.data.frame()

# Titling column name "symbol"
colnames(hallmark_emt_genes) <- c("symbol")

# EW24 EMT
ew24_emt_merge <- inner_join(all.EW24.tidy, hallmark_emt_genes, 
  by = join_by("symbol" == "symbol")) %>%
  dplyr::filter(adj.p.val <= 0.05 & log2FC >= 0.6)

ew24_emt_hp <- ew24_emt_merge %>% dplyr::select(3:4)

colnames(ew24_emt_hp) <- c("symbol","log2FC.EW24")

# SKNMC EMT
sknmc_emt_merge <- inner_join(all.SKNMC.tidy, hallmark_emt_genes, 
  by = join_by("symbol" == "symbol")) %>%
  dplyr::filter(adj.p.val <= 0.05 & log2FC >= 0.6)

sknmc_emt_hp <- sknmc_emt_merge %>% dplyr::select(3:4)

colnames(sknmc_emt_hp) <- c("symbol","log2FC.SKNMC")

# Merging SKNMC and EW24 upreg degs
gse7007_emt_merge <- inner_join(ew24_emt_hp, sknmc_emt_hp, 
  by = "symbol") %>%
  group_by(symbol) %>%
  filter(log2FC.SKNMC == max(log2FC.SKNMC)) %>% as.data.frame()

rownames(gse7007_emt_merge) <- gse7007_emt_merge$symbol

gse7007_emt_hp <- gse7007_emt_merge[,2:3]

## Heatmap
pheatmap(gse7007_emt_hp, fontsize = 7, angle = 45, cellheight = 6, cellwidth = 35, treeheight_row = 0, treeheight_col = 0,labels_col = c("EW24","SKNMC"),color=colorRampPalette(c("#3953A4","#4051A3","#5E4DA0","#7A459B","#903C97","#953A96","#963995","#A83192","#CD1580","#E61955","#ED1C28","#ED2024"))(100),main = paste("Hallmark EMT Genes (GSE7007)"))



```







```{r NOT USING}
results <- as.data.frame(fit2)
head(results)
write_csv(results, "gse7007_results.csv", na = "NA")

gse7007_results <- read_csv("/Users/emmacroushore/Downloads/gse7007_results.csv")

rs_reg <- gse7007_results %>%
  dplyr::select(11,12,1:3,24:26,36:38,42:43)

head(rs_reg)

qvals <- qvalue(rs_reg$p.value.EW24)
str(qvals)

rs_reg$F.q.value <- rs_reg$F.p.value %>%
  qvalue() %>% 
  .$qvalues

rs_reg$q.value.EW24 <- rs_reg$p.value.EW24 %>%
  qvalue() %>% 
  .$qvalues

rs_reg$q.value.SKNMC <- rs_reg$p.value.SKNMC %>%
  qvalue() %>% 
  .$qvalues

rs_reg$q.value.A673 <- rs_reg$p.value.A673 %>%
  qvalue() %>% 
  .$qvalues

rs_reg %>%
  filter(genes.Gene.symbol == "BCL2")

rs_min <- rs_reg %>%
  group_by(genes.Gene.symbol) %>%
  filter(F.p.value == min(F.p.value))

rs_min %>%
  filter(genes.Gene.symbol == "BCL2")

rs_tidy <- rs_min %>%
  pivot_longer(cols = c(9,3,10,4,11,5, 15:17), names_to = c(".value", "treatment"), names_pattern = "(^.+?).([^.]+$)") %>%
  arrange(F.p.value)

head(rs_tidy)

## Sig genes
ggplot(rs_tidy, aes(coefficients, -log10(p.value), color = ifelse(q.value <= 0.05, "qvalue < 0.05", "Unregulated"))) +
  geom_point(alpha = 0.5) +
  xlab("log2 Fold Change, shEWS-FLI1") +
  ylab("Significance (-log10(p value))") +
  theme_bw() +
  facet_wrap(~treatment)

rs_sig <- rs_tidy %>%
  dplyr::filter(q.value <= 0.05)

A673_sig <- rs_sig %>%
  dplyr::filter(treatment == "A673")
write_csv(A673_sig, "A673_sig.csv", na = "NA")

SKNMC_sig <- rs_sig %>%
  dplyr::filter(treatment == "SKNMC")
write_csv(SKNMC_sig, "SKNMC_sig.csv", na = "NA")

EW24_sig <- rs_sig %>%
  dplyr::filter(treatment == "EW24")
write_csv(EW24_sig, "EW24_sig.csv", na = "NA")


```

